### PR TITLE
[framework] inline casts on Element.widget getter to improve web performance

### PR DESCRIPTION
**Pull request by jonahwilliams**
This is an alternative to the proposal to make Element generic. After investigation, I discovered that the majority of the casting was actually unnecessary: here is an explanation:

```
class A {
  Object get foo;
}

class B extends A {
  String get foo => super.foo as String;
}

void main() {
  A a = B();
  Object bar = a.foo;
}
```

Consider the snippet above. Despite the fact that we only need an Object, as long as the subtype B is used there will be a cast performed on `B.foo`. Consider the Element/BuildContext API: the vast majority of the framework interacts with `Element.widget` or `BuildContext.widget` expecting only a `Widget` - but since all Element subclasses override the type of widget and use a cast, almost all accesses have a cast inserted even if the more specific Widget type is not required by the call site.

This leaves an alternative optimization - introduce a different getter with a cast and avoid overriding Element.widget. Furthermore, this presents a chance to optimize Element.widget by making the field non-nullable and using _NullWidget as a sentinel. This null check is fairly expensive in dart2js.

To benchmark this I used my favorite benchmark from https://github.com/flutter/flutter/pull/95596

On the web the average time of this benchmark decreased from 1845 ms to 1688 ms (~9%). On my Pixel 4 there was essentially no difference, which may just represent the lower cost of casts/null checks in AOT

The change to make Element generic was https://github.com/flutter/flutter/pull/97249 and would cause a breaking change in package:nested

### CODE REVIEW

File: packages/flutter/lib/src/widgets/framework.dart [hunk 1]
```diff
@@ -3233,8 +3233,8 @@ abstract class Element extends DiagnosticableTree implements BuildContext {
 
   /// The configuration for this element.
   @override
-  Widget get widget => _widget!;
-  Widget? _widget;
+  Widget get widget => _widget;
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/framework.dart:3236]: I thought we had documentation somewhere that was specifically telling people to override this with a getter that type casts to the right type. Obviously, that would need to be updated. But I can't find it, so maybe we don't?

Comment (goderbauer) [reply packages/flutter/lib/src/widgets/framework.dart:3236]: Ah! It's on RenderObjectElement: https://master-api.flutter.dev/flutter/widgets/RenderObjectElement-class.html

Can you update that doc there with the new practise?

File: packages/flutter/lib/src/widgets/framework.dart [hunk 2]
```diff
@@ -3233,8 +3233,8 @@ abstract class Element extends DiagnosticableTree implements BuildContext {
...
+  Widget _widget = const _NullWidget();
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/framework.dart:3237]: Maybe pull this out into a separate PR so we get separate benchmark runs for these two independent changes?

File: packages/flutter/lib/src/widgets/framework.dart (hunk 1 above)
Comment (goderbauer) [reply packages/flutter/lib/src/widgets/framework.dart:3236]: It may also be helpful to add a note to this doc string describing the pattern.

File: packages/flutter/lib/src/widgets/notification_listener.dart [hunk 3]
```diff
@@ -62,7 +62,7 @@ abstract class Notification {
   @mustCallSuper
   bool visitAncestor(Element element) {
     if (element is StatelessElement) {
-      final StatelessWidget widget = element.widget;
+      final StatelessWidget widget = element.widget as StatelessWidget;
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/notification_listener.dart:65]: I believe this cast can also be removed as follows since the code never relies on this being a StatelessWidget:

```suggestion
      final Widget widget = element.widget;
```

File: packages/flutter/lib/src/widgets/framework.dart [hunk 4]
```diff
@@ -5571,8 +5572,8 @@ abstract class RenderObjectElement extends Element {
   /// Creates an element that uses the given widget as its configuration.
   RenderObjectElement(RenderObjectWidget widget) : super(widget);
 
-  @override
-  RenderObjectWidget get widget => super.widget as RenderObjectWidget;
+  /// A typed override of [Element.widget].
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/framework.dart:5575]: Would be helpful to add a little color to this doc explaining when one should use `widget` and when `typedWidget` and why.

**Comment by goderbauer** [comment]
> On the web the average time of this benchmark decreased from 1845 ms to 1688 ms (~9%).

Nice!

> The change to make Element generic was https://github.com/flutter/flutter/pull/97249[](https://github.com/jonahwilliams) and would cause a breaking change in package:nested

Would the breakage be contained in package:nested and could simply be fixed there? Or would it also require changes to its customers?

File: packages/flutter/lib/src/widgets/framework.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/framework.dart:3236]: I think we could probably make this API mostly private too. I would need to update the doc with a better pattern

File: packages/flutter/lib/src/widgets/framework.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/framework.dart:3237]: Will do

File: packages/flutter/lib/src/widgets/notification_listener.dart (hunk 3 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/notification_listener.dart:65]: Ah nice

**Comment by jonahwilliams** [comment]
It could be fixed there, and I think it would be mostly self contained. The downside of making Element generic is that you can't make _widget non-nullable anymore, since it is typed `T` and we cannot create a const placeholder for an arbitrary T.

**Comment by Hixie** [comment]
> The downside of making Element generic is that you can't make _widget non-nullable anymore, since it is typed `T` and we cannot create a const placeholder for an arbitrary T.

could we type it `T extends Widget` and use your _NullWidget trick?

**Comment by jonahwilliams** [comment]
No, that wouldn't work since the Element subtypes would need to tighten the constraint, `StatelessElement<T extends StatelessWidget> extends ComponentElement<T>`, and `NullWidget` would no longer satisfy that

**Comment by Hixie** [comment]
Can we make it `late` and have a separate flag for when it's not set?

**Comment by jonahwilliams** [comment]
Unfortunately `late` also has terrible performance in dart2js....

**Comment by Hixie** [comment]
Is there a bug on that? It seems like `late` should be entirely free...

**Comment by jonahwilliams** [comment]
The nice part about making it non-nullable is that if you are using the API correctly there is no additional cost for accessing the field. The problem with late/nullable/a flag is that you need to insert a check on all accesses, and since JS is such a slower language than native machine code you pay an oversized effect.

**Comment by jonahwilliams** [comment]
I don't think its a bug, its just a cost of targeting a slow language

**Comment by Hixie** [comment]
In release builds there's no reason to add the test though right? That's what I mean by "free".

**Comment by jonahwilliams** [comment]
Actually I sort of take that back. For reference, in a native build on the same machine this benchmark takes 747 ms, which is about 2.2X as fast as the web version. If you assume the same lines of code run, and each line ran in about the same speed with the same optimization, then I would expect that eliminating the cast would have about half the impact on native.

But I don't see that at all, instead the needle doesn't move at all on a native build - which implies that `as` casts are much more expensive on the web.

**Comment by jonahwilliams** [comment]
@Hixie I believe dart2js is attempting to follow the spec which has specific kinds of behavior documented (i.e. LateInitializationError thrown) which is where the expense comes from.

**Comment by jonahwilliams** [comment]
Late issue is here: https://github.com/dart-lang/sdk/issues/43361

File: packages/flutter/lib/src/widgets/framework.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/framework.dart]: Removed

File: packages/flutter/lib/src/widgets/framework.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/framework.dart]: I've removed the public API

File: packages/flutter/lib/src/widgets/notification_listener.dart (hunk 3 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/notification_listener.dart]: Done

File: packages/flutter/lib/src/widgets/framework.dart (hunk 4 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/framework.dart:5573]: Since I removed the public API, I wonder what we should do here. We could warn about casting, but ideally we'd just fix that and delete the typedWidget private API later

**Comment by jonahwilliams** [comment]
I made all of the typedWidget getters private so there is no change to public API besides removing the more specific types from subclasses. I also removed the nullcheck - I did not observe a performance delta so I definitely overstated the impact of the nullcheck

File: packages/flutter/lib/src/widgets/notification_listener.dart (hunk 3 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/notification_listener.dart]: actually this whole notification API is sus, I need to look at this more later

File: packages/flutter/lib/src/widgets/framework.dart [hunk 5]
```diff
@@ -4359,7 +4359,7 @@ abstract class Element extends DiagnosticableTree implements BuildContext {
 
   /// A short, textual description of this element.
   @override
-  String toStringShort() => _widget?.toStringShort() ?? '${describeIdentity(this)}(DEFUNCT)';
+  String toStringShort() =>_widget?.toStringShort() ?? '${describeIdentity(this)}(DEFUNCT)';
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/framework.dart:4362]: typo

File: packages/flutter/lib/src/widgets/framework.dart [hunk 6]
```diff
@@ -4377,7 +4377,7 @@ abstract class Element extends DiagnosticableTree implements BuildContext {
     if (_lifecycleState != _ElementLifecycle.initial) {
       properties.add(ObjectFlagProperty<int>('depth', depth, ifNull: 'no depth'));
     }
-    properties.add(ObjectFlagProperty<Widget>('widget', _widget, ifNull: 'no widget'));
+    properties.add(ObjectFlagProperty<Widget>('widget',_widget, ifNull: 'no widget'));
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/framework.dart:4380]: typo

File: packages/flutter/lib/src/widgets/framework.dart (hunk 5 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/framework.dart]: fixed

File: packages/flutter/lib/src/widgets/framework.dart (hunk 6 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/framework.dart]: fixed

**Comment by Hixie** [comment]
This definitely removes some of my concerns.

File: packages/flutter/lib/src/widgets/framework.dart (hunk 1 above)
Comment (goderbauer) [reply packages/flutter/lib/src/widgets/framework.dart]: https://master-api.flutter.dev/flutter/widgets/RenderObjectElement-class.html still needs to be updated, though. It's telling people to do the casting pattern for new RenderObjectElements that we are now discouraging.

File: packages/flutter/lib/src/widgets/framework.dart (hunk 4 above)
Comment (goderbauer) [reply packages/flutter/lib/src/widgets/framework.dart]: What do you mean by "just fix that" - do you mean making it generic? Or do you see another way to avoid the private `typedWidget`s?

Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/framework.dart]: I mean ideally we would make casts faster, and then we wouldn't need to do the work around. I'll update the documentation to clarify the casting performance

**Comment by skia-gold** [comment]
Gold has detected about 2 new digest(s) on patchset 9.
View them at https://flutter-gold.skia.org/cl/github/97822

File: packages/flutter/lib/src/widgets/framework.dart (hunk 5 above)
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/framework.dart:4362]: nit: missing space after `=>`

File: packages/flutter/lib/src/widgets/framework.dart [hunk 7]
```diff
@@ -5433,19 +5421,24 @@ class InheritedElement extends ProxyElement {
...
-///   Foo get widget => super.widget as Foo;
-///
-///   @override
 ///   RenderFoo get renderObject => super.renderObject as RenderFoo;
 ///
+///   void _foo() {
+///     final Foo foo = widget as Foo;
+///   }
+///
+///   Foo get fooWidget => widget as Foo;
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/framework.dart:5441]: Maybe leave a comment above this line to make clear that this is an alternative approach to casting at the callsite to avoid the impression that its best practice to have this getter.

Comment (goderbauer) [reply packages/flutter/lib/src/widgets/framework.dart:5441]: Or maybe remove this line altogether. I think best practice should be to just cast at the callsite.

File: packages/flutter/lib/src/widgets/framework.dart [hunk 8]
```diff
@@ -5433,19 +5421,24 @@ class InheritedElement extends ProxyElement {
 /// ### Specializing the getters
 ///
 /// [RenderObjectElement] objects spend much of their time acting as
-/// intermediaries between their [widget] and their [renderObject]. To make this
-/// more tractable, most [RenderObjectElement] subclasses override these getters
-/// so that they return the specific type that the element expects, e.g.:
+/// intermediaries between their [widget] and their [renderObject]. It is
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/framework.dart:5424]: Can you also add a note to the docs of `Element.widget` explaining that it is totally expected to cast that at the callsite of the getter to the specific widget type of the element?

**Review by goderbauer — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/widgets/framework.dart (hunk 8 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/framework.dart:5424]: done, PTAL

File: packages/flutter/lib/src/widgets/framework.dart (hunk 7 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/framework.dart:5441]: done

File: packages/flutter/lib/src/widgets/framework.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/framework.dart:4380]: done

File: packages/flutter/lib/src/widgets/framework.dart (hunk 5 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/framework.dart:4362]: done

**Review by goderbauer — APPROVED** [review]
Still LGTM

File: packages/flutter/lib/src/widgets/framework.dart [hunk 9]
```diff
@@ -4021,7 +4028,7 @@ abstract class Element extends DiagnosticableTree implements BuildContext {
     assert(depth != null);
     assert(owner != null);
     // Use the private property to avoid a CastError during hot reload.
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/framework.dart:4030]: is this comment still accurate?

**Comment by Hixie** [comment]
test-exempt: code refactor with no semantic change

File: packages/flutter/lib/src/widgets/framework.dart (hunk 9 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/framework.dart:4030]: There may be Element subtypes that still have an override w/ cast outside of the framework
